app.py

```diff
@@ -46,158 +45,83 @@ def init_app(self):
-                                         path=path_on_disk, 
-                                         context=ctx, 
-                                         associated_entity=d.get("entity"))
-        except TankError, e:
-            self.log_error("Failed to launch an application for this published file: %s" % e)
+            if len(published_files) == 1:
+                published_file = self.execute_hook_method(
+                    "hook_get_published_file",
+                    "resolve_single_file",
+                    published_file_type=published_file_entity_type,
+                    published_file=published_files[0],
+                )
+            else:
+                published_file = self.execute_hook_method(
+                    "hook_get_published_file",
+                    "resolve_multiple_files",
+                    published_file_type=published_file_entity_type,
+                    published_files=published_files,
+                )
+        except (TankError, PublishPathNotDefinedError, PublishPathNotSupported) as e:
+            self.log_error(
+                "Failed to get a published file for %s %s: %s" % (
+                    published_file_entity_type,
+                    entity_ids[0],
+                    e))
             return
-        
-        if not launched:
-            # hook didn't know how to launch this
-            # just use std associated file launch
-            self.launch(path_on_disk)
 
-    def get_path_on_disk(self, published_file):
-        """
-        Retrieve, if possible, the path on disk of a PublishedFile.
+        # Finally, try to open the published file using launch hooks.
+        launch_hooks = self.get_setting("launch_publish_hooks")
+        errors = []
+        for launch_hook_expr in launch_hooks:
+            try:
+                launch_hook = self.create_hook_instance(launch_hook_expr)
+                launch_hook.execute(published_file=published_file)
+                return
+            except Exception as e:
+                message = "Failed to launch publish for %s with %s: %s" % (
+                    published_file,
+                    launch_hook_expr,
+                    e
+                )
+                self.log_debug(message)
+                errors.append(e)
+        self.log_error(
+            "Failed to Launch publish for %s: %s" % (
+                published_file, "\n".join(errors)
+            )
+        )
 
-        :param published_file: A PublishedFile dictionary.
-        :returns: The path on disk if found, otherwise ``None``.
+    def published_file(self, published_file_type, published_file_id):
         """
-        path_on_disk = published_file.get("path", {}).get("local_path")
-        if path_on_disk:
-            return path_on_disk
-        # If this PublishedFile came from a zero config publish, it will
-        # have a file URL rather than a local path.
-        url = published_file.get("path", {}).get("url")
```

Comment:
It seems the code dealing with urls has been lost?

Reply:
Nope, actually all this code was replaced by the `Hook.get_publish_path` which already deals with this logic:
https://github.com/shotgunsoftware/tk-core/blob/a98bbec19446244f4cfed8895aa926e0a34668d4/python/tank/util/shotgun/publish_resolve.py#L106